Document most of the remaining parts of the style system, make docs-by-default for the whole style system

emilio opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14819)
<!-- Reviewable:end -->

emilio committed:
style: remove ComputedStyle::do_cascade_property.

It's defined to be the same in both Servo and Gecko, so it's pointless right
now.

emilio committed:
style: Require docs in the less-straightforward part of the properties module.

emilio committed:
style: document the sink module.

emilio committed:
style: Document the restyle_damage module.

emilio committed:
style: Document the servo-specific module.

emilio committed:
style: Document the restyle hints code, and make it operate on `TElement`.

This removes the annoying constraint of having to provide the current state from
outside of the restyle hints code.

emilio committed:
style: Document and clean up the parser module.

emilio committed:
style: Document the `StyleContext` module.

emilio committed:
style: Document the scoped_tls module.

emilio committed:
style: Document the owning_handle module.

emilio committed:
style: Allow the lack of documentation for the media_queries module.

I'm rewriting all that stuff soon, so it's kind of pointless to document the
existing stuff right now.

emilio committed:
style: Document the keyframes module.

emilio committed:
style: Allow the `logical_geometry` module to be undocumented, for now.

emilio committed:
style: Allow the atomic_refcell module to be undocumented, for now.

Given Bobby is rewriting it as I write this.

emilio committed:
style: Document the custom_properties module.

emilio committed:
style: Allow the attr module to be without documentation, for now.

emilio committed:
style: Simplify the cascade function in the custom_properties module.

emilio committed:
style: Document the font_face module.

emilio committed:
style: Disallow new undocumented code by default for Servo.

Still a bunch of Gecko-only code to document.

emilio committed:
style: Remove a bunch of duplicated logic in add_stylesheet.

This is remminiscent of my recent @import refactor, and stayed there
incorrectly.

Note that it's the same code that lives in `update` (where it should live).

emilio committed:
style: Export fnv's FnvHashMap, instead of rewriting it.

emilio committed:
style: Document the stylist module.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/lib.rs, components/style/owning_handle.rs, components/style/custom_properties.rs, components/style/servo/mod.rs, components/style/data.rs, components/style/parser.rs, components/style/properties/gecko.mako.rs, components/style/restyle_hints.rs, components/style/properties/helpers.mako.rs, components/style/font_face.rs, components/style/servo/restyle_damage.rs, components/style/scoped_tls.rs, components/style/keyframes.rs, components/style/sink.rs, components/style/selector_parser.rs, components/style/properties/properties.mako.rs, components/style/context.rs, components/style/stylist.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

emilio commented:
BTW, remaining parts are:

 * The `longhands` and `shorthands` modules. My idea for those, given it's mostly parsing code, would be to require just a spec link per property.

 * The parts that are being rewritten (`atomic_refcell` and `media_queries`), because I don't want to waste too much time.

 * The `attr` module.

 * The `gecko` bits. Probably not too hard to do.

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 770b4ee with merge 81795a9...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4974)

emilio committed:
style: Document the StyleComplexColor sugar.

emilio committed:
style: Document the RefPtr sugar module.

stshine commented:
You are so amazing.

Manishearth commented:
Reviewed 2 of 2 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r8.
Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/style/properties/properties.mako.rs, line 486 at r2](https://reviewable.io:443/reviews/servo/servo/14819#-K_Rgh6OsCIkWcIRddyP:-K_Rgh6OsCIkWcIRddyQ:b-6zuhvd) ([raw file](https://github.com/servo/servo/blob/a1a6c27d4ce17cace7fefff5f53771939c2a39c9/components/style/properties/properties.mako.rs#L486)):*
> ```Rust
>     }
> 
>     /// Trie to serialize the given declarations as this shorthand, or return an
> ```

"try"

"return an error if writing to the stream fails. In case the declarations do not map to a shorthand, this will not write anything but will return a success result."

---

*[components/style/properties/properties.mako.rs, line 671 at r2](https://reviewable.io:443/reviews/servo/servo/14819#-K_RhDirzebrJ2Lz4B4-:-K_RhDirzebrJ2Lz4B40:bm0wn7g) ([raw file](https://github.com/servo/servo/blob/a1a6c27d4ce17cace7fefff5f53771939c2a39c9/components/style/properties/properties.mako.rs#L671)):*
> ```Rust
> 
> impl<'a> PropertyDeclarationId<'a> {
>     /// Whether a given declaration id is either the same as `other`,  or a
> ```

double space

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14819)*
<!-- Sent from Reviewable.io -->

emilio committed:
style: Document and make clearer the ownership helpers.

emilio committed:
style: Document nsTArray sugar module.

emilio committed:
style: Document the bindings module, allow missing docs in autogenerated files.

emilio committed:
style: Describe the sugar module.

emilio committed:
style: Document sugar for nsStyleAutoArray.

emilio committed:
style: Document the sugar for `nsStyleCoord`.

emilio committed:
style: Document nsCSSShadowArray sugar.

emilio committed:
style: Document Gecko's selector parser.

emilio committed:
style: Document the nsCOMPtr sugar.

emilio committed:
style: Ditch GeckoStyleCoordHelpers, and implement directly on nsStyleCoord.

emilio committed:
style: Document Gecko conversions module.

emilio committed:
style: Document Gecko's data module.

emilio committed:
style: Describe the gecko module.

emilio committed:
style: Document Gecko's restyle-damage code.

emilio committed:
style: Document Gecko's snapshot code.

emilio committed:
style: Document Gecko's traversal code.

Actually, I think most of the functions in DOMTraversalContext deserve a nice
comment, but that's probably fine as a followup, I want to land the require-docs
thing now.

emilio committed:
style: Document the remaining public parts of Gecko's values module.

emilio committed:
style: Document Gecko's wrapper module.

emilio committed:
style: Document gecko_string_cache.

emilio committed:
style: Require documentation also for a Gecko build.

emilio commented:
Addressed your nits, went ahead with the Gecko docs too, and fixed the build error that was because I changed the stylist's `viewport_constraints` to return `Option<&ViewportConstraints>` instead of `&Option<ViewportConstraints>`, but I assumed it was only used in style and thus `Ctrl-C`d the build.

I'd appreciate a fast-ish review time, since pretty much any patch that checks in undocumented code in the style system will make me grow this patch.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit caebd5f with merge 80a45c5...

emilio commented:
cc @heycam @SimonSapin @upsuper and @bzbarsky, which aren't mentioned automatically by bors and haven't been assigned nor reviewed anything.

If you have any objection to this, please speak before landing. Also, review if you feel like it :)

I'm aware the docs for some methods are probably not as extensive as they should, but it's five AM and oh well...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4688), [arm32](http://build.servo.org/builders/arm32/builds/4857), [arm64](http://build.servo.org/builders/arm64/builds/4888), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4979), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1656), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1662), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4956), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4995), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1661), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1698), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/118), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/119)

Manishearth commented:
Reviewed 37 of 37 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/style/parser.rs, line 47 at r9](https://reviewable.io:443/reviews/servo/servo/14819#-K_Tuq39pJvzJ1kvrveM:-K_Tuq39pJvzJ1kvrveN:b-6co8it) ([raw file](https://github.com/servo/servo/blob/caebd5f98501e4563a9998e9ac0a5983e298f4c6/components/style/parser.rs#L47)):*
> ```Rust
> /// The data that the parser needs from outside in order to parse a stylesheet.
> pub struct ParserContext<'a> {
>     /// The `Origin` of the stylesheet.
> ```

Might be better to be explicit here -- "The `Origin` of a stylesheet -- whether it is a user, author, or user-agent sheet"

(`Origin` is easily confused with URL origins)

---

*[components/style/gecko_bindings/sugar/ns_com_ptr.rs, line 5 at r9](https://reviewable.io:443/reviews/servo/servo/14819#-K_Ttk1xFow3PqI4B5sE:-K_Ttk1xFow3PqI4B5sF:b2bpqex) ([raw file](https://github.com/servo/servo/blob/caebd5f98501e4563a9998e9ac0a5983e298f4c6/components/style/gecko_bindings/sugar/ns_com_ptr.rs#L5)):*
> ```Rust
>  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
> 
> //! Litle helpers for `nsCOMPtr`.
> ```

speling

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14819)*
<!-- Sent from Reviewable.io -->

Manishearth commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r10, 1 of 1 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14819#-:-K_TyyjzbxnB5dVmTfrE:bnfp4nl)*
<!-- Sent from Reviewable.io -->

emilio commented:
On Mon, Jan 02, 2017 at 03:25:14AM -0800, Manish Goregaokar wrote:
> *[components/style/parser.rs, line 47 at r9](https://reviewable.io:443/reviews/servo/servo/14819#-K_Tuq39pJvzJ1kvrveM:-K_Tuq39pJvzJ1kvrveN:b-6co8it) ([raw file](https://github.com/servo/servo/blob/caebd5f98501e4563a9998e9ac0a5983e298f4c6/components/style/parser.rs#L47)):*
> > ```Rust
> > /// The data that the parser needs from outside in order to parse a stylesheet.
> > pub struct ParserContext<'a> {
> >     /// The `Origin` of the stylesheet.
> > ```
> 
> Might be better to be explicit here -- "The `Origin` of a stylesheet -- whether it is a user, author, or user-agent sheet"
> 
> (`Origin` is easily confused with URL origins)

Good call, done :)

> *[components/style/gecko_bindings/sugar/ns_com_ptr.rs, line 5 at r9](https://reviewable.io:443/reviews/servo/servo/14819#-K_Ttk1xFow3PqI4B5sE:-K_Ttk1xFow3PqI4B5sF:b2bpqex) ([raw file](https://github.com/servo/servo/blob/caebd5f98501e4563a9998e9ac0a5983e298f4c6/components/style/gecko_bindings/sugar/ns_com_ptr.rs#L5)):*
> > ```Rust
> >  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
> > 
> > //! Litle helpers for `nsCOMPtr`.
> > ```
> 
> speling

Good catch.

emilio commented:
@bors-servo r=Manishearth

Thanks for the review Manish :)

bors-servo commented:
:pushpin: Commit a8fdf5b has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth a8fdf5b6b9cf615c7ad18d01871491f3e0ca2562 -->

bors-servo commented:
:hourglass: Testing commit a8fdf5b with merge 8784732...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4692), [arm32](http://build.servo.org/builders/arm32/builds/4861), [arm64](http://build.servo.org/builders/arm64/builds/4892), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4983), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1660), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1666), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4960), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4999), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1665), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1702), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/122), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/123)

upsuper commented on components/style/gecko_bindings/sugar/style_complex_color.rs:
```diff
@@ -17,6 +19,7 @@ impl From<nscolor> for StyleComplexColor {
 }
 
 impl StyleComplexColor {
+    /// Create a `StyleComplexColor` value that represents `currentColor`.
```
[`currentcolor` is speced all-lowercased now](https://drafts.csswg.org/css-color-4/#currentcolor-color).
